Process crash recovery stuck

### Version

0.9.23

### Current behavior

A network error makes the realtime indexing process to end up in a recovery loop that fails after multiple attempts to handle reorg.

<details><summary>Logs</summary>

```
9:52:42 PM INFO  database   Dropped index 'new_ttls_domain_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'new_ttls_domain_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'interface_changed_resolver_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'interface_changed_resolver_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'authorisation_changed_resolver_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'authorisation_changed_resolver_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'version_changed_resolver_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'version_changed_resolver_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Reverted 0 unfinalized operations from 'new_ttls'

9:52:42 PM INFO  database   Reverted 0 unfinalized operations from 'fuses_set'

9:52:42 PM INFO  database   Reverted 0 unfinalized operations from 'expiry_extended'

9:52:42 PM INFO  database   Reverted 0 unfinalized operations from 'abi_changed'

9:52:42 PM INFO  database   Reverted 0 unfinalized operations from 'pubkey_changed'

9:52:42 PM INFO  database   Reverted 0 unfinalized operations from 'contenthash_changed'

9:52:42 PM INFO  database   Reverted 0 unfinalized operations from 'interface_changed'

9:52:42 PM INFO  database   Reverted 0 unfinalized operations from 'authorisation_changed'

9:52:42 PM INFO  database   Reverted 0 unfinalized operations from 'version_changed'

9:52:42 PM INFO  database   Reverted 1 unfinalized operations from 'transfers'

9:52:42 PM INFO  database   Reverted 1 unfinalized operations from 'name_unwrapped'

9:52:42 PM INFO  database   Reverted 3 unfinalized operations from 'name_renewed'

9:52:42 PM INFO  database   Reverted 4 unfinalized operations from 'name_wrapped'

9:52:42 PM INFO  database   Reverted 4 unfinalized operations from 'text_changed'

9:52:42 PM INFO  database   Dropped index 'wrapped_transfers_domain_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'wrapped_transfers_domain_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_wrapped_domain_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_wrapped_domain_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_unwrapped_domain_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_unwrapped_domain_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'fuses_set_domain_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'fuses_set_domain_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'expiry_extended_domain_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'expiry_extended_domain_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_registered_registration_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_registered_registration_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_renewed_registration_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_renewed_registration_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_transferred_registration_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_transferred_registration_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'addr_changed_resolver_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'addr_changed_resolver_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'multicoin_addr_changed_resolver_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'multicoin_addr_changed_resolver_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_changed_resolver_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'name_changed_resolver_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'abi_changed_resolver_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'abi_changed_resolver_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'pubkey_changed_resolver_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'pubkey_changed_resolver_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'text_changed_resolver_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'text_changed_resolver_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'contenthash_changed_resolver_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:52:42 PM INFO  database   Dropped index 'contenthash_changed_resolver_id_id_index' in schema '524e4f1a-ca69-4232-b85f-31e7943f3f74'

9:53:46 PM WARN  database   Failed 'revert' database method after '10' attempts (id=bb07eb8a)

error: current transaction is aborted, commands ignored until end of transaction block

    at /app/node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/client.js:526:17

    at processTicksAndRejections (node:internal/process/task_queues:105:5)

    at file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:1298:30

    at async Promise.all (index 0)

    at Object.wrap (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:529:26)

    at Object.revert (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:1292:7)

    at file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:1175:15

    at NodePgSession.transaction (file:///app/node_modules/.pnpm/drizzle-orm@0.39.3_@electric-sql+pglite@0.2.13_@opentelemetry+api@1.7.0_kysely@0.26.3_pg@8.11.3/node_modules/src/node-postgres/session.ts:193:19)

    at Object.wrap (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:529:26)

    at run (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/bin/utils/run.ts:41:29)

9:53:46 PM WARN  database   Failed 'recoverCheckpoint' database method after '10' attempts (id=bfb236ba)

error: null value in column "domain_id" of relation "wrapped_domains" violates not-null constraint

    at /app/node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/client.js:526:17

    at processTicksAndRejections (node:internal/process/task_queues:105:5)

    at file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:1351:17

    at async Promise.all (index 4)

    at Object.wrap (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:529:26)

    at Object.revert (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:1292:7)

    at file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:1175:15

    at NodePgSession.transaction (file:///app/node_modules/.pnpm/drizzle-orm@0.39.3_@electric-sql+pglite@0.2.13_@opentelemetry+api@1.7.0_kysely@0.26.3_pg@8.11.3/node_modules/src/node-postgres/session.ts:193:19)

    at Object.wrap (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:529:26)

    at run (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/bin/utils/run.ts:41:29)

9:53:46 PM ERROR process    Caught unhandledRejection event

error: null value in column "domain_id" of relation "wrapped_domains" violates not-null constraint

    at /app/node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/client.js:526:17

    at processTicksAndRejections (node:internal/process/task_queues:105:5)

    at file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:1351:17

    at async Promise.all (index 4)

    at Object.wrap (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:529:26)

    at Object.revert (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:1292:7)

    at file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:1175:15

    at NodePgSession.transaction (file:///app/node_modules/.pnpm/drizzle-orm@0.39.3_@electric-sql+pglite@0.2.13_@opentelemetry+api@1.7.0_kysely@0.26.3_pg@8.11.3/node_modules/src/node-postgres/session.ts:193:19)

    at Object.wrap (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/database/index.ts:529:26)

    at run (file:///app/node_modules/.pnpm/ponder@0.9.23_@opentelemetry+api@1.7.0_@types+node@20.17.14_@types+react@19.0.10_hono@4.6.17__x432yxsydyoxz433hikz5mk6da/node_modules/ponder/src/bin/utils/run.ts:41:29)

9:53:46 PM WARN  process    Received unhandledRejection, starting shutdown sequence

/app/apps/ensindexer:

 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @ensnode/ensindexer@0.0.1 start: `ponder start "--schema" "524e4f1a-ca69-4232-b85f-31e7943f3f74"`

Exit status 1
```
</details>

### Expected behavior

A network error should allow ponder process to fail and restart from latest checkpoint to continue realtime indexing.

### Steps to reproduce

Provoke a process restart while indexing in realtime phase (i.e.indexing function with unhandled network error).

### Link to repository

https://github.com/namehash/ensnode/tree/main/apps/ensindexer

### Anything else?

My plan is to handle runtime exception so I could avoid process restarts. However, if the ponder app process restarts for any other reason, it should be able to continue indexing from the last checkpoint.

--- a/.changeset/slow-donkeys-love.md
+++ b/.changeset/slow-donkeys-love.md
@@ -1,0 +1,5 @@
+---
+"ponder": patch
+---
+
+Fixed a bug that would cause crash recovery and reorg handling to not work for some applications.

--- a/.changeset/slow-donkeys-love.md
+++ b/.changeset/slow-donkeys-love.md
@@ -2,4 +2,4 @@
 "ponder": patch
 ---
 
-Fixed a bug that would cause crash recovery and reorg handling to not work for some applications.
+Fixed a bug that would cause crash recovery and reorg handling to error for some apps.

--- a/packages/core/src/database/index.test.ts
+++ b/packages/core/src/database/index.test.ts
@@ -720,13 +720,20 @@ test("revert()", async (context) => {
     .values({ address: "0x0000000000000000000000000000000000000001" });
 
   await database.complete({
+    checkpoint: createCheckpoint({ chainId: 1n, blockNumber: 10n }),
+    db: database.qb.drizzle,
+  });
+
+  await indexingStore.delete(account, { address: zeroAddress });
+
+  await database.complete({
     checkpoint: createCheckpoint({ chainId: 1n, blockNumber: 11n }),
     db: database.qb.drizzle,
   });
 
   await database.qb.drizzle.transaction(async (tx) => {
     await database.revert({
-      checkpoint: createCheckpoint({ chainId: 1n, blockNumber: 10n }),
+      checkpoint: createCheckpoint({ chainId: 1n, blockNumber: 9n }),
       tx,
     });
   });

--- a/packages/core/src/database/index.test.ts
+++ b/packages/core/src/database/index.test.ts
@@ -678,6 +678,18 @@ test("complete()", async (context) => {
 });
 
 test("revert()", async (context) => {
+  const test = onchainTable(
+    "Test",
+    (p) => ({
+      a: p.integer("A").notNull(),
+      b: p.integer("B").notNull(),
+      c: p.integer("C"),
+    }),
+    (table) => ({
+      pk: primaryKey({ columns: [table.a, table.b] }),
+    }),
+  );
+
   const database = await createDatabase({
     common: context.common,
     namespace: "public",
@@ -685,8 +697,8 @@ test("revert()", async (context) => {
       databaseConfig: context.databaseConfig,
     },
     schemaBuild: {
-      schema: { account },
-      statements: buildSchema({ schema: { account } }).statements,
+      schema: { account, test },
+      statements: buildSchema({ schema: { account, test } }).statements,
     },
   });
 
@@ -698,7 +710,7 @@ test("revert()", async (context) => {
 
   const indexingStore = createRealtimeIndexingStore({
     common: context.common,
-    schemaBuild: { schema: { account } },
+    schemaBuild: { schema: { account, test } },
     database,
   });
 
@@ -742,6 +754,32 @@ test("revert()", async (context) => {
 
   expect(rows).toHaveLength(1);
   expect(rows[0]).toStrictEqual({ address: zeroAddress, balance: 10n });
+
+  await indexingStore.insert(test).values({ a: 1, b: 1 });
+
+  await database.complete({
+    checkpoint: createCheckpoint({ chainId: 1n, blockNumber: 11n }),
+    db: database.qb.drizzle,
+  });
+
+  await indexingStore.update(test, { a: 1, b: 1 }).set({ c: 1 });
+
+  await database.complete({
+    checkpoint: createCheckpoint({ chainId: 1n, blockNumber: 12n }),
+    db: database.qb.drizzle,
+  });
+
+  await database.qb.drizzle.transaction(async (tx) => {
+    await database.revert({
+      checkpoint: createCheckpoint({ chainId: 1n, blockNumber: 11n }),
+      tx,
+    });
+  });
+
+  const rows2 = await database.qb.drizzle.select().from(test);
+
+  expect(rows2).toHaveLength(1);
+  expect(rows2[0]).toStrictEqual({ a: 1, b: 1, c: null });
 
   await context.common.shutdown.kill();
 });

--- a/packages/core/src/database/index.test.ts
+++ b/packages/core/src/database/index.test.ts
@@ -678,18 +678,6 @@ test("complete()", async (context) => {
 });
 
 test("revert()", async (context) => {
-  const test = onchainTable(
-    "Test",
-    (p) => ({
-      a: p.integer("A").notNull(),
-      b: p.integer("B").notNull(),
-      c: p.integer("C"),
-    }),
-    (table) => ({
-      pk: primaryKey({ columns: [table.a, table.b] }),
-    }),
-  );
-
   const database = await createDatabase({
     common: context.common,
     namespace: "public",
@@ -697,8 +685,8 @@ test("revert()", async (context) => {
       databaseConfig: context.databaseConfig,
     },
     schemaBuild: {
-      schema: { account, test },
-      statements: buildSchema({ schema: { account, test } }).statements,
+      schema: { account },
+      statements: buildSchema({ schema: { account } }).statements,
     },
   });
 
@@ -710,7 +698,7 @@ test("revert()", async (context) => {
 
   const indexingStore = createRealtimeIndexingStore({
     common: context.common,
-    schemaBuild: { schema: { account, test } },
+    schemaBuild: { schema: { account } },
     database,
   });
 
@@ -755,6 +743,46 @@ test("revert()", async (context) => {
   expect(rows).toHaveLength(1);
   expect(rows[0]).toStrictEqual({ address: zeroAddress, balance: 10n });
 
+  await context.common.shutdown.kill();
+});
+
+test("revert() with composite primary key", async (context) => {
+  const test = onchainTable(
+    "Test",
+    (p) => ({
+      a: p.integer("A").notNull(),
+      b: p.integer("B").notNull(),
+      c: p.integer("C"),
+    }),
+    (table) => ({
+      pk: primaryKey({ columns: [table.a, table.b] }),
+    }),
+  );
+
+  const database = await createDatabase({
+    common: context.common,
+    namespace: "public",
+    preBuild: {
+      databaseConfig: context.databaseConfig,
+    },
+    schemaBuild: {
+      schema: { test },
+      statements: buildSchema({ schema: { test } }).statements,
+    },
+  });
+
+  await database.migrate({ buildId: "abc" });
+
+  // setup tables, reorg tables, and metadata checkpoint
+
+  await database.createTriggers();
+
+  const indexingStore = createRealtimeIndexingStore({
+    common: context.common,
+    schemaBuild: { schema: { test } },
+    database,
+  });
+
   await indexingStore.insert(test).values({ a: 1, b: 1 });
 
   await database.complete({
@@ -776,10 +804,10 @@ test("revert()", async (context) => {
     });
   });
 
-  const rows2 = await database.qb.drizzle.select().from(test);
+  const rows = await database.qb.drizzle.select().from(test);
 
-  expect(rows2).toHaveLength(1);
-  expect(rows2[0]).toStrictEqual({ a: 1, b: 1, c: null });
+  expect(rows).toHaveLength(1);
+  expect(rows[0]).toStrictEqual({ a: 1, b: 1, c: null });
 
   await context.common.shutdown.kill();
 });

--- a/packages/core/src/database/index.ts
+++ b/packages/core/src/database/index.ts
@@ -1293,7 +1293,7 @@ FOR EACH ROW EXECUTE FUNCTION "${namespace}".${getTableNames(table).triggerFn};
           tables.map(async (table) => {
             const primaryKeyColumns = getPrimaryKeyColumns(table);
 
-            await tx.execute(
+            const result = await tx.execute(
               sql.raw(`
 WITH reverted1 AS (
   DELETE FROM "${namespace}"."${getTableName(getReorgTable(table))}"
@@ -1309,11 +1309,13 @@ WITH reverted1 AS (
   ON ${primaryKeyColumns.map(({ sql }) => `reverted2."${sql}" = reverted1."${sql}"`).join("AND ")}
   AND reverted2.operation_id = reverted1.operation_id
 ), inserted AS (
-  DELETE FROM "${namespace}"."${getTableName(table)}"
-  WHERE ${primaryKeyColumns.map(({ sql }) => `"${sql}"`).join(", ")} IN (
-    SELECT ${primaryKeyColumns.map(({ sql }) => `"${sql}"`).join(",")} FROM reverted3
-    WHERE operation = 0
+  DELETE FROM "${namespace}"."${getTableName(table)}" as t
+  WHERE EXISTS (
+    SELECT * FROM reverted3
+    WHERE ${primaryKeyColumns.map(({ sql }) => `t."${sql}" = reverted3."${sql}"`).join("AND ")}
+    AND OPERATION = 0
   )
+  RETURNING *
 ), updated_or_deleted AS (
   INSERT INTO  "${namespace}"."${getTableName(table)}"
   SELECT ${Object.values(getTableColumns(table))
@@ -1328,12 +1330,15 @@ WITH reverted1 AS (
           `"${getColumnCasing(column, "snake_case")}" = EXCLUDED."${getColumnCasing(column, "snake_case")}"`,
       )
       .join(", ")}
-) SELECT;`),
+  RETURNING *
+) SELECT (SELECT COUNT(*) FROM updated_or_deleted) + (SELECT COUNT(*) FROM inserted) as count;
+`),
             );
 
             common.logger.info({
               service: "database",
-              msg: `Reverted unfinalized operations from '${getTableName(table)}'`,
+              // @ts-ignore
+              msg: `Reverted ${result.rows[0]!.count} unfinalized operations from '${getTableName(table)}'`,
             });
           }),
         ),

--- a/packages/core/src/database/index.ts
+++ b/packages/core/src/database/index.ts
@@ -30,7 +30,6 @@ import { wait } from "@/utils/wait.js";
 import type { PGlite } from "@electric-sql/pglite";
 import {
   type TableConfig,
-  and,
   eq,
   getTableColumns,
   getTableName,
@@ -1289,77 +1288,58 @@ FOR EACH ROW EXECUTE FUNCTION "${namespace}".${getTableNames(table).triggerFn};
       });
     },
     async revert({ checkpoint, tx }) {
-      await this.wrap({ method: "revert", includeTraceLogs: true }, () =>
+      await this.record({ method: "revert", includeTraceLogs: true }, () =>
         Promise.all(
           tables.map(async (table) => {
             const primaryKeyColumns = getPrimaryKeyColumns(table);
 
-            // @ts-ignore
-            const { rows } = await tx.execute<Schema>(
-              sql.raw(
-                `DELETE FROM "${namespace}"."${getTableName(getReorgTable(table))}" WHERE checkpoint > '${checkpoint}' RETURNING *`,
-              ),
+            await tx.execute(
+              sql.raw(`
+WITH reverted1 AS (
+  DELETE FROM "${namespace}"."${getTableName(getReorgTable(table))}"
+  WHERE checkpoint > '${checkpoint}' RETURNING *
+), reverted2 AS (
+  SELECT ${primaryKeyColumns.map(({ sql }) => `"${sql}"`).join(", ")}, MIN(operation_id) AS operation_id FROM reverted1
+  GROUP BY ${primaryKeyColumns.map(({ sql }) => `"${sql}"`).join(", ")}
+), reverted3 AS (
+  SELECT ${Object.values(getTableColumns(table))
+    .map((column) => `reverted1."${getColumnCasing(column, "snake_case")}"`)
+    .join(", ")}, reverted1.operation FROM reverted2
+  INNER JOIN reverted1
+  ON ${primaryKeyColumns.map(({ sql }) => `reverted2."${sql}" = reverted1."${sql}"`).join("AND ")}
+  AND reverted2.operation_id = reverted1.operation_id
+), inserted AS (
+  DELETE FROM "${namespace}"."${getTableName(table)}"
+  WHERE ${primaryKeyColumns.map(({ sql }) => `"${sql}"`).join(", ")} IN (
+    SELECT ${primaryKeyColumns.map(({ sql }) => `"${sql}"`).join(",")} FROM reverted3
+    WHERE operation = 0
+  )
+), updated AS (
+  INSERT INTO  "${namespace}"."${getTableName(table)}"
+  SELECT ${Object.values(getTableColumns(table))
+    .map((column) => `"${getColumnCasing(column, "snake_case")}"`)
+    .join(", ")} FROM reverted3
+  WHERE operation = 1
+  ON CONFLICT (${primaryKeyColumns.map(({ sql }) => `"${sql}"`).join(", ")})
+  DO UPDATE SET
+    ${Object.values(getTableColumns(table))
+      .map(
+        (column) =>
+          `"${getColumnCasing(column, "snake_case")}" = EXCLUDED."${getColumnCasing(column, "snake_case")}"`,
+      )
+      .join(", ")}
+), deleted AS (
+  INSERT INTO "${namespace}"."${getTableName(table)}"
+  SELECT ${Object.values(getTableColumns(table))
+    .map((column) => `"${getColumnCasing(column, "snake_case")}"`)
+    .join(", ")} FROM reverted3
+  WHERE operation = 2
+) SELECT * FROM reverted3;`),
             );
-
-            const reversed = rows.sort(
-              // @ts-ignore
-              (a, b) => b.operation_id - a.operation_id,
-            );
-
-            // undo operation
-            for (const log of reversed) {
-              if (log.operation === 0) {
-                // create
-
-                await tx.delete(table).where(
-                  and(
-                    ...primaryKeyColumns.map(({ js, sql }) =>
-                      // @ts-ignore
-                      eq(table[js]!, log[sql]),
-                    ),
-                  ),
-                );
-              } else if (log.operation === 1) {
-                // update
-
-                // @ts-ignore
-                log.operation_id = undefined;
-                // @ts-ignore
-                log.checkpoint = undefined;
-                // @ts-ignore
-                log.operation = undefined;
-                await tx
-                  .update(table)
-                  .set(log)
-                  .where(
-                    and(
-                      ...primaryKeyColumns.map(({ js, sql }) =>
-                        // @ts-ignore
-                        eq(table[js]!, log[sql]),
-                      ),
-                    ),
-                  );
-              } else {
-                // delete
-
-                // @ts-ignore
-                log.operation_id = undefined;
-                // @ts-ignore
-                log.checkpoint = undefined;
-                // @ts-ignore
-                log.operation = undefined;
-                await tx
-                  .insert(table)
-                  .values(log)
-                  .onConflictDoNothing({
-                    target: primaryKeyColumns.map(({ js }) => table[js]!),
-                  });
-              }
-            }
 
             common.logger.info({
               service: "database",
-              msg: `Reverted ${rows.length} unfinalized operations from '${getTableName(table)}'`,
+              msg: `Reverted unfinalized operations from '${getTableName(table)}'`,
             });
           }),
         ),

--- a/packages/core/src/database/index.ts
+++ b/packages/core/src/database/index.ts
@@ -1314,12 +1314,12 @@ WITH reverted1 AS (
     SELECT ${primaryKeyColumns.map(({ sql }) => `"${sql}"`).join(",")} FROM reverted3
     WHERE operation = 0
   )
-), updated AS (
+), updated_or_deleted AS (
   INSERT INTO  "${namespace}"."${getTableName(table)}"
   SELECT ${Object.values(getTableColumns(table))
     .map((column) => `"${getColumnCasing(column, "snake_case")}"`)
     .join(", ")} FROM reverted3
-  WHERE operation = 1
+  WHERE operation = 1 OR operation = 2
   ON CONFLICT (${primaryKeyColumns.map(({ sql }) => `"${sql}"`).join(", ")})
   DO UPDATE SET
     ${Object.values(getTableColumns(table))
@@ -1328,13 +1328,7 @@ WITH reverted1 AS (
           `"${getColumnCasing(column, "snake_case")}" = EXCLUDED."${getColumnCasing(column, "snake_case")}"`,
       )
       .join(", ")}
-), deleted AS (
-  INSERT INTO "${namespace}"."${getTableName(table)}"
-  SELECT ${Object.values(getTableColumns(table))
-    .map((column) => `"${getColumnCasing(column, "snake_case")}"`)
-    .join(", ")} FROM reverted3
-  WHERE operation = 2
-) SELECT * FROM reverted3;`),
+) SELECT;`),
             );
 
             common.logger.info({

--- a/packages/core/src/database/index.ts
+++ b/packages/core/src/database/index.ts
@@ -1331,7 +1331,7 @@ WITH reverted1 AS (
       )
       .join(", ")}
   RETURNING *
-) SELECT (SELECT COUNT(*) FROM updated_or_deleted) + (SELECT COUNT(*) FROM inserted) as count;
+) SELECT COUNT(*) FROM reverted3 as count;
 `),
             );
 

--- a/packages/core/src/database/index.ts
+++ b/packages/core/src/database/index.ts
@@ -1331,7 +1331,7 @@ WITH reverted1 AS (
       )
       .join(", ")}
   RETURNING *
-) SELECT COUNT(*) FROM reverted3 as count;
+) SELECT COUNT(*) FROM reverted1 as count;
 `),
             );
 

